<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

## Lab: Cleaning Rock Song Data

_Author: Naumaan Nayyar_

---


In [156]:
import pandas as pd
import numpy as np 
import seaborn as sns

%matplotlib inline

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns

%matplotlib inline

### 1. Load `rock.csv` and do an initial examination of its data columns.

In [157]:
rockfile = "./datasets/rock.csv"

In [10]:
# Load the data.

In [11]:
# Look at the information regarding its columns.

### 2.  Clean up the column names.

Let's clean up the column names. There are two ways we can accomplish this:

#### 2.A Change the column names when you import the data using `pd.read_csv()`.

Notice that, when passing `names=[..A LIST OF STRING..]` with a number of columns that matches the number of strings in the passed list, you replace the column names.

NOTE: When you create custom column names, the first row of the `.csv` already represents a header. It is important to tell `pandas` to skip that row. The `skiprows=1` keyword argument to `read_csv()` will tell `pandas` to skip the first row.

In [150]:
# Change the column names when loading the '.csv':
rock = pd.read_csv("./datasets/rock.csv")

In [151]:
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Song Clean    2230 non-null   object
 1   ARTIST CLEAN  2230 non-null   object
 2   Release Year  1653 non-null   object
 3   COMBINED      2230 non-null   object
 4   First?        2230 non-null   int64 
 5   Year?         2230 non-null   int64 
 6   PlayCount     2230 non-null   int64 
 7   F*G           2230 non-null   int64 
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


#### 2.B Change column names using the `.rename()` function.

The `.rename()` function takes an argument, `columns=name_dict`, in which `name_dict` is a dictionary containing the original column names as keys and the new column names as values.

In [152]:
# Change the column names using the `.rename()` function.
rock.rename(columns = {"song":"Song Clean","artist":"ARTIST CLEAN","release":"Release Year","combine":"COMBINED","first":"First?","year":"Year?","play":"PlayCount","fg":"F*G"}, inplace=True)

In [153]:
rock.rename(columns={'beer_servings':'beer', 'wine_servings':'wine'}, inplace=True)

In [160]:
cols = ["song","artist","release","combine","first","year","play","fg"]
rock = pd.read_csv(rockfile, names = cols)

In [22]:
rock.head()

song        artist       release  \
0              Song Clean  ARTIST CLEAN  Release Year   
1        Caught Up in You   .38 Special          1982   
2            Fantasy Girl   .38 Special           NaN   
3         Hold On Loosely   .38 Special          1981   
4  Rockin' Into the Night   .38 Special          1980   

                                 combine   first   year       play   fg  
0                               COMBINED  First?  Year?  PlayCount  F*G  
1        Caught Up in You by .38 Special       1      1         82   82  
2            Fantasy Girl by .38 Special       1      0          3    0  
3         Hold On Loosely by .38 Special       1      1         85   85  
4  Rockin' Into the Night by .38 Special       1      1         18   18

#### 2.C Reassigning the `.columns` attribute of a DataFrame.

You can also just reassign the `.columns` attribute to a list of strings containing the new column names. 

The only caveat with reassigning `.columns` is that you have to reassign all of the column names at once. You can't partially replace a value by working on `.columns` directly. You have to reassign `.columns` with a list of equal length. 

In [ ]:
# Replace the column names by reassigning the `.columns` attribute.


### 3. Subsetting data where null values exist.

We have mixed `str` and `NaN` values in the `release` column. `NaN` stands for "not a number" and is the way `pandas` handles "nulls" or nonexistent data. We can use the `.isnull()` method of a Series to find null values.

Print the header of the data subset to where the `release` column is null values.

In [31]:
# Show records where df['release'] is null
mask = rock['release'].isnull()
rock[mask].head()

song       artist release  \
2             Fantasy Girl  .38 Special     NaN   
11   Baby, Please Don't Go        AC/DC     NaN   
14  CAN'T STOP ROCK'N'ROLL        AC/DC     NaN   
17        Girls Got Rhythm        AC/DC     NaN   
25         Let's Get It Up        AC/DC     NaN   

                            combine first year play fg  
2       Fantasy Girl by .38 Special     1    0    3  0  
11   Baby, Please Don't Go by AC/DC     1    0    1  0  
14  CAN'T STOP ROCK'N'ROLL by AC/DC     1    0    5  0  
17        Girls Got Rhythm by AC/DC     1    0   24  0  
25         Let's Get It Up by AC/DC     1    0    4  0

### 4. Update slices of your DataFrame based on mask selection/slices.

In many scenarios, we want to upate values in our DataFrame according to criteria. Let's say we wanted to set all of the null values in `release` to 0.

With newer versions of `pandas`, in order to manipulate data in the original DataFrame, we have to use `.loc` while performing reassignment using a mask and an index.

For example, the following won't always work:
```python
df[row_mask]['column_name'] = new_value
```

The best way to accomplish the same task is:
```python
df.loc[row_mask, 'column_name'] = new_value
```

For multiple column assignment, you would use:
```python
df.loc[row_mask, ['col_1', 'col_2', 'col_3']] = new_value
```

#### 4.A Let's try it out. Make all of the null values in `release` 0.

In [32]:
# Replace release nulls with 0
rock.loc[mask,['release']] = 0

#### 4.B Verify that `release` contains no null values.

In [35]:
# A:
rock['release'].isnull().sum()

0

### 5. Ensure that the data types of the columns make sense. 

Verifying column data types is a critical part of data munging. If columns have the wrong data type, then there is usually corrupted or incorrect data in some of the observations.

#### 5.A Look at the data types for the columns. Are any incorrect given what the data represents?

In [36]:
# A:
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2231 entries, 0 to 2230
Data columns (total 8 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   song     2231 non-null   object
 1   artist   2231 non-null   object
 2   release  2231 non-null   object
 3   combine  2231 non-null   object
 4   first    2231 non-null   object
 5   year     2231 non-null   object
 6   play     2231 non-null   object
 7   fg       2231 non-null   object
dtypes: object(8)
memory usage: 139.6+ KB


In [99]:
column_names = ['song', 'artist', 'release', 'song_artist', 'first', 'year', 'playcount', 'fg']
rock = pd.read_csv(rockfile, names=column_names, skiprows=1)

rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   song         2230 non-null   object
 1   artist       2230 non-null   object
 2   release      1653 non-null   object
 3   song_artist  2230 non-null   object
 4   first        2230 non-null   int64 
 5   year         2230 non-null   int64 
 6   playcount    2230 non-null   int64 
 7   fg           2230 non-null   int64 
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


### 6. Investigate and clean up the `release` column.

The `release` column is a string data type when it should be an integer.

#### 6.A Figure out what value(s) are causing the `release` column to be encoded as a string instead of an integer.

In [100]:
# A:
rock.release.unique()

array(['1982', nan, '1981', '1980', '1975', '2000', '2002', '1992',
       '1985', '1993', '1976', '1995', '1979', '1984', '1977', '1990',
       '1986', '1974', '2014', '1987', '1973', '2001', '1989', '1997',
       '1971', '1972', '1994', '1970', '1966', '1965', '1983', '1955',
       '1978', '1969', '1999', '1968', '1988', '1962', '2007', '1967',
       '1958', '1071', '1996', '1991', '2005', '2011', '2004', '2012',
       '2003', '1998', '2008', '1964', '2013', '2006', 'SONGFACTS.COM',
       '1963', '1961'], dtype=object)

#### 6.B Look at the rows in which there is incorrect data in the `release` column.

In [106]:
# A:
mask = rock['release'] == "SONGFACTS.COM"

C:\Users\zhang\anaconda3\lib\site-packages\pandas\core\ops\array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


In [107]:
rock.loc[mask, 'release'] = np.nan
rock['release'] = rock['release'].map(lambda x: float(x))

#### 6.C. Clean up the data. Normally we may replace the offending data with null np.nan values, however we previously converted all of the nan values in the release column to zeros so we might as well continue with the same practice. Replacing with 0 (or nan) will allow us to convert the column to numeric.

In [108]:
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   song         2230 non-null   object 
 1   artist       2230 non-null   object 
 2   release      1652 non-null   float64
 3   song_artist  2230 non-null   object 
 4   first        2230 non-null   int64  
 5   year         2230 non-null   int64  
 6   playcount    2230 non-null   int64  
 7   fg           2230 non-null   int64  
dtypes: float64(1), int64(4), object(3)
memory usage: 139.5+ KB


### 7. Get summary statistics for the `release` column using the `.describe()` function.

Now that the `release` column is finally a numeric data type, we can apply the `.describe()` function.  

#### 7.A Print out the summary stats for the `release` column. What is the earliest and latest release date?

In [109]:
# A:
rock.release.describe()

count    1652.000000
mean     1978.019976
std        24.191247
min      1071.000000
25%      1971.000000
50%      1977.000000
75%      1984.000000
max      2014.000000
Name: release, dtype: float64

#### 7.B Based on the summary statistics, is there anything else wrong with the `release` column? 

_Looking at the DataFrame that contains the year 1071, we can see that the year was probably corrupted and should be replaced with something else if possible._

In [133]:
rock[rock['release']==1070]

Empty DataFrame
Columns: [song, artist, release, song_artist, first, year, playcount, fg]
Index: []

### 8. Make changes and investigate using custom functions with `.apply()`.

Let's say we want to traverse every single row in our data set and apply a function to that row.

#### 8.A Write a function that will take a row of a DataFrame and print out the song, artist, and whether or not the release date is < 1970.


In [119]:
# A:
def rock_new(row):
    print('--------------------------------')
    print(row['song'], row['artist'], row['release'], '< 1970?:', row['release'] < 1970)

#### 8.B Using the `.apply()` function, apply the function you wrote to the first four rows of the DataFrame.

You will need to tell the `apply` function to operate row by row. Setting the keyword argument as `axis=1` indicates that the function should be applied to each row individually.

In [120]:
# A:
rock.head(4).apply(rock_new, axis=1)

--------------------------------
Caught Up in You .38 Special 1982.0 < 1970?: False
--------------------------------
Fantasy Girl .38 Special nan < 1970?: False
--------------------------------
Hold On Loosely .38 Special 1981.0 < 1970?: False
--------------------------------
Rockin' Into the Night .38 Special 1980.0 < 1970?: False


0    None
1    None
2    None
3    None
dtype: object

You'll notice that there will be a final output Series of `None` values. The `.apply()` function, if a return value is not specified, will return a Series of `None` values (similar to how the default return for Python functions is `None` when a return statement is not specified).

### 9. Write a function that converts cells in a DataFrame to float and otherwise replaces them with `np.nan`.

If applied to our data, it would keep only the numeric information and otherwise input null values.

Recall that the try-except syntax in Python is a great way to try something and take another action if the initial step fails:

```python
try:
    Perform some action.
except:
   Perform some other action if the first failed with an error.
```

#### 9.A Write the function that takes a column and converts all of its values to float if possible and `np.nan` otherwise. The return value should be the converted Series.

In [121]:
# A:
def converter_helper(value):
    try:
        return float(value)
    except:
        return np.nan

def convert_to_float(column):
    column = column.map(converter_helper)
    return column

In [134]:
column_names = ['song', 'artist', 'release', 'first', 'year', 'play', 'fg']
rock = pd.read_csv(rockfile, names=column_names, skiprows=1)

rock.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2230 entries, Caught Up in You to WAITIN' FOR THE BUS/JESUS JUST LEFT CHICAGO
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   song     2230 non-null   object
 1   artist   1653 non-null   object
 2   release  2230 non-null   object
 3   first    2230 non-null   int64 
 4   year     2230 non-null   int64 
 5   play     2230 non-null   int64 
 6   fg       2230 non-null   int64 
dtypes: int64(4), object(3)
memory usage: 139.4+ KB


#### 9.B Try your function out on the rock song data and ensure the output is what you expected.


In [135]:
# A:
rock.apply(convert_to_float).head(10)

song  artist  release  first  year  play    fg
Caught Up in You         NaN  1982.0      NaN    1.0   1.0  82.0  82.0
Fantasy Girl             NaN     NaN      NaN    1.0   0.0   3.0   0.0
Hold On Loosely          NaN  1981.0      NaN    1.0   1.0  85.0  85.0
Rockin' Into the Night   NaN  1980.0      NaN    1.0   1.0  18.0  18.0
Art For Arts Sake        NaN  1975.0      NaN    1.0   1.0   1.0   1.0
Kryptonite               NaN  2000.0      NaN    1.0   1.0  13.0  13.0
Loser                    NaN  2000.0      NaN    1.0   1.0   1.0   1.0
When I'm Gone            NaN  2002.0      NaN    1.0   1.0   6.0   6.0
What's Up?               NaN  1992.0      NaN    1.0   1.0   3.0   3.0
Take On Me               NaN  1985.0      NaN    1.0   1.0   1.0   1.0

#### 9.C Describe the new float-only DataFrame.

In [136]:
# A:
rock_new = rock.apply(convert_to_float)

In [137]:
rock_new.describe()

song       artist  release   first         year         play  \
count   0.0  1652.000000      0.0  2230.0  2230.000000  2230.000000   
mean    NaN  1978.019976      NaN     1.0     0.741256    16.872646   
std     NaN    24.191247      NaN     0.0     0.438043    25.302972   
min     NaN  1071.000000      NaN     1.0     0.000000     0.000000   
25%     NaN  1971.000000      NaN     1.0     0.000000     1.000000   
50%     NaN  1977.000000      NaN     1.0     1.000000     4.000000   
75%     NaN  1984.000000      NaN     1.0     1.000000    21.000000   
max     NaN  2014.000000      NaN     1.0     1.000000   142.000000   

                fg  
count  2230.000000  
mean     15.048430  
std      25.288366  
min       0.000000  
25%       0.000000  
50%       3.000000  
75%      18.000000  
max     142.000000